In [4]:
import pandas as pd 
df = pd.read_csv(r"C:\Data Analytics\Scraping data\fahasascraper\fahasascraper\books_data.csv")

# Extract cột Series/Episode từ BookName

In [5]:

def extract_series_episode(row):
    book_name = row['book_name']
    book_name_origi = row['book_name_origi']
    
    if " - Tập" in book_name:
        series, episode = book_name.split(" - Tập", 1)
        return pd.Series([series.strip(), f"Tập {episode.strip()}"])
    else:
        return pd.Series([book_name_origi, None])

# Apply the function to create 'series' and 'episode' columns
df[['series', 'episode']] = df.apply(extract_series_episode, axis=1)

# Extract cột Sell_method (tặng kèm hay không) và Gift_method (tặng kèm cái gì)

In [4]:
# Function to determine sell_method based on "Tặng ..." in book_name
def determine_sell_method(book_name):
    if "Tặng" in book_name:
        return "Tặng " + book_name.split("Tặng", 1)[1].strip()
    else:
        return "Không tặng kèm"

# Apply the function to create the 'sell_method' column
df['sell_method'] = df['book_name'].apply(determine_sell_method)

In [4]:
# Rename 'sell_method' to 'gift_method'
df1.rename(columns={'sell_method': 'gift_method'}, inplace=True)

# Create a new column 'gift_or_not'
df1['gift_or_not'] = df1['gift_method'].apply(lambda x: "Không tặng kèm" if x.lower() == "không tặng kèm" else "Có tặng kèm")


             product_id                                          book_name  \
0              9.79E+12            Lịch Sử 11 (Chân Trời Sáng Tạo) (Chuẩn)   
1              9.79E+12  Tiếng Anh 3 - Sách Bài Tập (Global Success) (T...   
2              9.79E+12           38 Bức Thư Rockefeller Viết Cho Con Trai   
3              9.79E+12                        Địa Lí 11 (Kết Nối) (Chuẩn)   
4              9.79E+12     Khoa Học Tự Nhiên 6 (Kết Nối Tri Thức) (Chuẩn)   
...                 ...                                                ...   
27030          8.94E+12  Bài Tập Tuần Toán 2 - Tập 2 (Biên Soạn Theo Ch...   
27031          8.94E+12  Truyện Cổ Tích Kinh Điển - Ba Chú Heo Con (Sti...   
27032          8.94E+12                  Endurance - Hành Trình Không Gian   
27033          8.93E+12  Bé Tập Kể Chuyện - Nàng Tiên Bóng Đêm (Tái Bản...   
27034  combo-1607202409  Combo Manga - Doraemon - Tuyển Tập Theo Chủ Đề...   

                                                  series  \
0  

## Tiếp tục chỉnh sửa lần thứ 2

In [2]:
df1 = pd.read_csv(r"C:\Data Analytics\Scraping data\fahasascraper\fahasascraper\books_data_ver2.csv")
df1

product_id                                          book_name  \
0              9.79E+12            Lịch Sử 11 (Chân Trời Sáng Tạo) (Chuẩn)   
1              9.79E+12  Tiếng Anh 3 - Sách Bài Tập (Global Success) (T...   
2              9.79E+12           38 Bức Thư Rockefeller Viết Cho Con Trai   
3              9.79E+12                        Địa Lí 11 (Kết Nối) (Chuẩn)   
4              9.79E+12     Khoa Học Tự Nhiên 6 (Kết Nối Tri Thức) (Chuẩn)   
...                 ...                                                ...   
27030          8.94E+12  Bài Tập Tuần Toán 2 - Tập 2 (Biên Soạn Theo Ch...   
27031          8.94E+12  Truyện Cổ Tích Kinh Điển - Ba Chú Heo Con (Sti...   
27032          8.94E+12                  Endurance - Hành Trình Không Gian   
27033          8.93E+12  Bé Tập Kể Chuyện - Nàng Tiên Bóng Đêm (Tái Bản...   
27034  combo-1607202409  Combo Manga - Doraemon - Tuyển Tập Theo Chủ Đề...   

                                                  series  \
0                Lịch Sử 11 (Chân Trời Sáng Tạo) (Chuẩn)   
1      Tiếng Anh 3 - Sách Bài Tập (Global Success) (T...   
2               38 Bức Thư Rockefeller Viết Cho Con Trai   
3                            Địa Lí 11 (Kết Nối) (Chuẩn)   
4         Khoa Học Tự Nhiên 6 (Kết Nối Tri Thức) (Chuẩn)   
...                                                  ...   
27030                                Bài Tập Tuần Toán 2   
27031  Truyện Cổ Tích Kinh Điển - Ba Chú Heo Con (Sti...   
27032                  Endurance - Hành Trình Không Gian   
27033  Bé Tập Kể Chuyện - Nàng Tiên Bóng Đêm (Tái Bản...   
27034  Combo Manga - Doraemon - Tuyển Tập Theo Chủ Đề...   

                                                 episode     gift_or_not  \
0                                                    NaN  Không tặng kèm   
1                                                    NaN  Không tặng kèm   
2                                                    NaN  Không tặng kèm   
3                                                    NaN  Không tặng kèm   
4                                                    NaN  Không tặng kèm   
...                                                  ...             ...   
27030  Tập 2 (Biên Soạn Theo Chương Trình Giáo Dục Ph...  Không tặng kèm   
27031                                                NaN  Không tặng kèm   
27032                                                NaN  Không tặng kèm   
27033                                                NaN  Không tặng kèm   
27034                                                NaN  Không tặng kèm   

          gift_method             category_1           category_2      price  \
0      Không tặng kèm  Giáo khoa - Tham khảo       Sách Giáo Khoa   15.000 đ   
1      Không tặng kèm  Giáo khoa - Tham khảo       Sách Giáo Khoa   62.000 đ   
2      Không tặng kèm  Tâm lý - Kỹ năng sống         Kỹ năng sống   69.000 đ   
3      Không tặng kèm  Giáo khoa - Tham khảo       Sách Giáo Khoa   23.000 đ   
4      Không tặng kèm  Giáo khoa - Tham khảo       Sách Giáo Khoa   27.000 đ   
...               ...                    ...                  ...        ...   
27030  Không tặng kèm  Giáo khoa - Tham khảo       Sách Tham Khảo   41.250 đ   
27031  Không tặng kèm              Thiếu nhi     Truyện Thiếu Nhi   38.400 đ   
27032  Không tặng kèm         Tiểu Sử Hồi Ký  Câu Chuyện Cuộc Đời  186.750 đ   
27033  Không tặng kèm              Thiếu nhi     Truyện Thiếu Nhi   12.750 đ   
27034  Không tặng kèm          Manga - Comic                Manga  336.140 đ   

       old_price  ... quantity_sold  \
0            NaN  ...          1.1k   
1            NaN  ...           292   
2      138.000 đ  ...          3.5k   
3            NaN  ...           280   
4            NaN  ...            98   
...          ...  ...           ...   
27030   55.000 đ  ...            51   
27031   48.000 đ  ...            52   
27032  249.000 đ  ...            51   
27033   15.000 đ  ...            51   
27034  350.000 đ  ...            52   

            

## đổi tên các cột

In [2]:
df1 = df1.rename(columns={"Năm Xuất Bản" : "nam_xuat_ban",
                          "Ngôn Ngữ" : "ngon_ngu",
                          "Trọng Lượng":"trong_luong",
                          "Kích Thước":"kich_thuoc",
                          "Số Trang":"so_trang",
                          "Loại Bìa":"loai_bia"})

NameError: name 'df1' is not defined

## sửa cột price, old price

In [4]:
df1= df1.rename(columns={"price":"new_price"})


In [5]:
df1["old_price"] = df1["old_price"].fillna(df1["new_price"])

In [8]:
import pandas as pd
import re

# Function to remove 'đ', spaces, commas, and other unwanted characters, then return as a numeric value
def clean_currency(value):
    if isinstance(value, str):
        # Use regex to remove all non-numeric characters (except for the decimal point)
        cleaned_value = re.sub(r'[^\d.]', '', value)
        return pd.to_numeric(cleaned_value, errors='coerce')
    return value

# Apply the function to both 'new_price' and 'old_price' columns
df1['new_price'] = df1['new_price'].apply(clean_currency)
df1['old_price'] = df1['old_price'].apply(clean_currency)
df1['new_price'] = df1['new_price'] * 1000
df1['old_price'] = df1['old_price'] * 1000

## sửa cột quantity sold

In [10]:
import pandas as pd

# Function to replace 'k', 'k+', and '+' with '000' or clean up the value
def replace_k(value):
    if isinstance(value, str):
        # Remove the '+' sign if it exists
        value = value.replace('+', '')
        
        # Check for 'k' or 'k+' and multiply by 1000
        if 'k' in value:
            return float(value.replace('k', '')) * 1000
    return value

# Apply the function to 'quantity_sold' column
df1['quantity_sold'] = df1['quantity_sold'].apply(replace_k)


## fix product_id và add book_type column

In [11]:
def product_id_fix(row,index):
    if 'combo' not in row ['book_name'].lower():
        return index +1 
    else :
        return row['product_id']
mask = ~df1['book_name'].str.contains('combo',case=False, na=False)
df1['product_id']=df1.apply(lambda row: product_id_fix(row,df1.index.get_loc(row.name)),axis=1)

In [12]:
import pandas as pd

# Function to classify book type as 'single' or 'combo'
def book_type_column(value):
    # Check if the value is numeric (i.e., int or float)
    if isinstance(value, (int, float)):
        return "single"
    else:
        return "combo"

# Apply the function to create a new column 'book_type'
df1['book_type'] = df1['product_id'].apply(book_type_column)


# Sửa cột nam_xuat_ban

In [17]:
import pandas as pd
import re

# Function to clean 'nam_xuat_ban' based on your requirements
def clean_nam_xuat_ban(value):
    # Check if the value is a valid year in YYYY format
    match = re.match(r'^\d{4}$', str(value))
    
    if match:
        year = int(match.group())
        # Remove rows where the year is greater than 2024
        if year > 2024:
            return None  # This will be used to drop the row later
        return year
    else:
        # If it's not a 4-digit year, replace it with 2024
        return 2024

# Apply the cleaning function to 'nam_xuat_ban' column
df3['nam_xuat_ban'] = df3['nam_xuat_ban'].apply(clean_nam_xuat_ban)

# Drop rows where 'nam_xuat_ban' is None (invalid years)
df3 = df3.dropna(subset=['nam_xuat_ban'])


# Sửa data lần thứ 3 

In [7]:
import pandas as pd 
df3 = pd.read_csv(r"C:\Data Analytics\Scraping data\fahasascraper\fahasascraper\books_data_ver3.csv")

In [37]:
df3

product_id  \
0                                       525   
1                                       605   
2                 combo-dstsmg12-dstsln5-qt   
3                                       642   
4      combo-onepiece103ba-onepiece103gh-qt   
...                                     ...   
27033                                 26409   
27034                                 26480   
27035     combo-9786045833261-9786045896754   
27036                                 26756   
27037                                 26875   

                                               book_name  \
0                             Bạch Tuyết Tóc Đỏ - Tập 20   
1      Chu Du Ẩm Thực Tại Dị Giới Với Kĩ Năng Không T...   
2      Combo Sách Dược Sư Tự Sự - Tập 12 (Manga) + Tậ...   
3      [Light Novel] Dược Sư Tự Sự - Tập 5 - Tặng Kèm...   
4      Combo Manga - One Piece - Tập 103 - Bản Bìa Áo...   
...                                                  ...   
27033  Hành Trang Cho Bé Tự Tin Vào Lớp 1 - Tập Tô Th...   
27034  Ghép Hình - Kể Chuyện: Nàng Công Chúa Ngủ Tron...   
27035  Combo Sự Phát Triển Của Tư Tưởng Ở Việt Nam Từ...   
27036  Hộp Quà 1/6 - Độ Tuổi 3-6: Hộp Hè Khám Phá - T...   
27037  Baby Book - Chúc Bé Ngủ Ngon (Túi 4 Tập) (Tái ...   

                                                  series  \
0                                      Bạch Tuyết Tóc Đỏ   
1      Chu Du Ẩm Thực Tại Dị Giới Với Kĩ Năng Không T...   
2                               Combo Sách Dược Sư Tự Sự   
3                            [Light Novel] Dược Sư Tự Sự   
4                                Combo Manga - One Piece   
...                                                  ...   
27033                 Hành Trang Cho Bé Tự Tin Vào Lớp 1   
27034  Ghép Hình - Kể Chuyện: Nàng Công Chúa Ngủ Tron...   
27035  Combo Sự Phát Triển Của Tư Tưởng Ở Việt Nam Từ...   
27036  Hộp Quà 1/6 - Độ Tuổi 3-6: Hộp Hè Khám Phá - T...   
27037  Baby Book - Chúc Bé Ngủ Ngon (Túi 4 Tập) (Tái ...   

                                                 episode     gift_or_not  \
0                                                 Tập 20  Không tặng kèm   
1                       Tập 10 - Tặng Kèm Special Jacket     Có tặng kèm   
2      Tập 12 (Manga) + Tập 5 (Light Novel) (Bộ 2 Cuố...     Có tặng kèm   
3                      Tập 5 - Tặng Kèm Bookmark + Sổ Xé     Có tặng kèm   
4      Tập 103 - Bản Bìa Áo + Limited Edition (Bộ 2 C...     Có tặng kèm   
...                                                  ...             ...   
27033                       Tập Tô Thông Minh (Bộ 2 Tập)  Không tặng kèm   
27034                                                NaN  Không tặng kèm   
27035                                                NaN  Không tặng kèm   
27036                                                NaN  Không tặng kèm   
27037                                                NaN  Không tặng kèm   

                                             gift_method  \
0                                         Không tặng kèm   
1                                Tặng Kèm Special Jacket   
2      Tặng Kèm Standee Ivory + Bìa Áo 2 Mặt + Bookma...   
3                              Tặng Kèm Bookmark + Sổ Xé   
4                    Tặng Kèm Bìa Áo Hologram + Postcard   
...                                                  ...   
27033                                     Không tặng kèm   
27034                                     Không tặng kèm   
27035                                     Không tặng kèm   
27036                                     Không tặng kèm   
27037                                     Không tặng kèm   

                  category_1                        category_2  new_price  \
0              Manga - Comic                             Manga      28500   
1              Manga - Comic                             Manga      38000   
2              Manga - Comic                             Manga     157150   
3                    Văn học                       Light Novel     112500   
4      

# sửa lại cột tên của nhà xuất bản

In [16]:
nxb_data = pd.read_csv(r"C:\Data Analytics\fahasa_project\nhà_xuất_bản_data.csv")
nxb_data

nha_xuat_ban  \
0                                 Hoa Học Trò   
1        Báo Sinh Viên Việt Nam - Hoa Học Trò   
2        Báo Sinh Viên Việt Nam - Hoa Học Trò   
3                              Báo Tiền Phong   
4                  Cambridge University Press   
..                                        ...   
288  Tài Nguyên Môi Trường Và Bản Đồ Việt Nam   
289                            Tân Việt Books   
290                         Từ Điển Bách Khoa   
291      Vietnam News Agency Publishing House   
292          Vietnam Women's Publishing House   

                            nha_xuat_ban_name  
0        Báo Sinh Viên Việt Nam - Hoa Học Trò  
1        Báo Sinh Viên Việt Nam - Hoa Học Trò  
2        Báo Sinh Viên Việt Nam - Hoa Học Trò  
3                              Báo Tiền Phong  
4                        Cambridge University  
..                                        ...  
288  Tài Nguyên Môi Trường Và Bản Đồ Việt Nam  
289                            Tân Việt Books  
290                     NXB Từ Điển Bách Khoa  
291      Vietnam News Agency Publishing House  
292          Vietnam Women's Publishing House  

[293 rows x 2 columns]

# Left join df3 with nxb_data (Từ 291 Nhà xuất bản => 80)

In [ ]:
# Perform the left join on 'nha_xuat_ban'
df3_merged = pd.merge(df3, nxb_data, on='nha_xuat_ban', how='left')

# Drop the old 'nha_xuat_ban' column
df3_merged = df3_merged.drop(columns=['nha_xuat_ban'])

# Print all unique values of 'nha_xuat_ban_name'
unique_nha_xuat_ban_name = df3_merged['nha_xuat_ban_name'].unique()
unique_nha_xuat_ban_name

In [28]:
unique_nha_xuat_ban_count = df3_merged['nha_xuat_ban_name'].nunique()
unique_nha_xuat_ban_count

83

# sửa lại cột book_type

In [1]:
import pandas as pd
df = pd.read_csv(r"C:\Data Analytics\Scraping data\fahasascraper\fahasascraper\books_data_ver3.csv")

In [2]:
# Count unique values in the book_type column
book_type_counts = df['book_type'].value_counts()

# Display the result
print(book_type_counts)


book_type
single    26407
combo       631
Name: count, dtype: int64


In [7]:
import re


# Update book_type to "combo" based on conditions
df['book_type'] = df.apply(
    lambda row: 'combo' if re.search(r"Bộ\s[1-9]", row['book_name']) or 'combo' in row['product_id'] 
    else row['book_type'], 
    axis=1
)



In [8]:
book_type_counts = df['book_type'].value_counts()

# Display the result
print(book_type_counts)


book_type
single    25875
combo      1163
Name: count, dtype: int64


# Save datafile lại

In [5]:
# Save the updated DataFrame to a new CSV file with utf-8-sig encoding
df.to_csv(r"C:\Data Analytics\Scraping data\fahasascraper\fahasascraper\books_data_ver3.csv", index=False, encoding='utf-8-sig')
